src/codegen/traverse.lisp

```diff
@@ -250,9 +252,47 @@ other nodes, then it would be inappropriate to also define an action
               (fire-action ':traverse (class-name (class-of current-node)) actions args)
               (fire-action ':after    (class-name (class-of current-node)) actions args)
               (fire-action ':after    'node                                actions args))))
-      (let ((*traverse* #'current-traverse))
+      (let ((*traverse* (lambda (node &rest args)
+                          (funcall *traverse-hook*
+                                   (lambda ()
+                                     (apply #'current-traverse node args))
+                                   node))))
         (apply *traverse* initial-node initial-args)))))
 
+(defun identity-hook (thunk node)
+  "Default traversal hook function, which doesn't do any extra operation."
+  (declare (ignore node)) (funcall thunk))
+
+(defvar *traverse-hook* #'identity-hook)
+
+(defun traverse-with-hook (hook node action-list &rest args)
```

Comment:
Add documentation to the `traverse` function saying something like: "See `traverse-with-hook` for a more customizable variant." or something like that.